internal/providers/terraform/parser.go

```diff
@@ -720,19 +720,18 @@ func addressModulePart(addr string) string {
 }
 
 func getModuleNames(addr string) []string {
-	r := regexp.MustCompile(`module\.([^\.\[]*)`)
-	matches := r.FindAllStringSubmatch(addressModulePart(addr), -1)
-
-	if matches == nil {
-		return []string{}
+	modulesPart := strings.TrimSuffix(addressModulePart(addr), ".")
+	if strings.HasPrefix(modulesPart, "module.") {
+		modulesPart = fmt.Sprintf(".%s", modulesPart)
 	}
 
-	n := make([]string, 0, len(matches))
-	for _, m := range matches {
-		n = append(n, m[1])
+	result := []string{}
+	for _, p := range strings.Split(modulesPart, ".module.") {
+		if p != "" {
+			result = append(result, p)
+		}
 	}
-
-	return n
+	return result
```

Comment:
@hugorut I'm 99% sure there should be a better way to do it, please advise. After all my debugging it looks like the parsed module name keeps the index, so I changed this method.

@aliscott Do you happen to remember why you excluded the indexes from module names? I'm worried my change can break something else, so more context could help :) Thanks.

Reply:
@vdmgolub this does not appear to be the solution to the actual problem. The module config in an actual plan JSON does not appear to have the keys/indexs. For example take the following terraform:

```
provider "aws" {
  ...
}

module "test" {
  count  = 2
  source = "./modules/test"
}
```

the `configuration.root_module` has the following in the plan JSON:

```
    "root_module": {
      "module_calls": {
        "test": {
           ...
          "module": {...}
         }
      } 
```

wheras when we run the HCL provider it builds the following json output:

```
    "root_module": {
      "module_calls": {
        "test[0]": {
          ...
          "module": {...}
        },
        "test[1]": {
          ...
           "module": {...}
        }
     }
```

See how it is incorrectly adding the keys to the configuration section.

Hence the fix relies not on altering this function but in the `HCLProvider.marshalModule` functionality.

Reply:
Thanks! I missed comparing the JSON plans 🤦. Here's the update.